beacon-chain/state/state-native/setters_block.go

```diff
@@ -43,24 +50,33 @@ func (b *BeaconState) SetBlockRoots(val [][]byte) error {
 // UpdateBlockRootAtIndex for the beacon state. Updates the block root
 // at a specific index to a new value.
 func (b *BeaconState) UpdateBlockRootAtIndex(idx uint64, blockRoot [32]byte) error {
-	if uint64(len(b.blockRoots)) <= idx {
-		return fmt.Errorf("invalid index provided %d", idx)
-	}
-	b.lock.Lock()
-	defer b.lock.Unlock()
+	if features.Get().EnableExperimentalState {
+		if err := b.blockRootsMultiValue.UpdateAt(b, idx, blockRoot); err != nil {
+			return errors.Wrap(err, "could not update block roots")
+		}
+	} else {
+		if uint64(len(b.blockRoots)) <= idx {
+			return errors.Wrapf(consensus_types.ErrOutOfBounds, "block root index %d does not exist", idx)
+		}
 
-	r := b.blockRoots
-	if ref := b.sharedFieldReferences[types.BlockRoots]; ref.Refs() > 1 {
-		// Copy elements in underlying array by reference.
-		roots := *b.blockRoots
-		rootsCopy := roots
-		r = &rootsCopy
-		ref.MinusRef()
-		b.sharedFieldReferences[types.BlockRoots] = stateutil.NewRef(1)
+		b.lock.Lock()
```

Comment:
you need to lock at the start of the function

Reply:
I simplified the locking to the simplest pattern possible